labgrid/driver/sshdriver.py

```diff
@@ -222,6 +224,82 @@ def interact(self, cmd=None):
         )
         return sub.wait()
 
+    @Driver.check_active
+    @contextlib.contextmanager
+    def forward_local_port(self, remoteport, localport=None):
+        """Forward a local port to a remote port on the target
+
+        A context manager that keeps a local port forwarded to a remote port as
+        long as the context remains valid. A connection can be made to the
+        returned port on localhost and it will be forwarded to the remote port
+        on the target device
+
+        usage:
+            with ssh.forward_local_port(8080) as localport:
+                # Use localhost:localport here to connect to port 8080 on the
+                # target
+
+        returns:
+        localport
+        """
+        if not self._check_keepalive():
+            raise ExecutionError("Keepalive no longer running")
+
+        if localport is None:
+            localport = get_free_port()
+
+        complete_cmd = ["ssh",
+                "-L", "%d:localhost:%d" % (localport, remoteport),
+                "-o", "ControlPath={}".format(self.control.replace('%', '%%')),
+                "-N",
+                "_",
+        ]
+        self.logger.debug("Running command: %s", complete_cmd)
+        sub = subprocess.Popen(complete_cmd)
+        try:
+            yield localport
+        finally:
+            sub.kill()
+            sub.wait()
+
+    @Driver.check_active
+    @contextlib.contextmanager
+    def forward_remote_port(self, remoteport, localport=None):
```

Comment:
Random local target port doesn't seem to be useful? And probing a random port on the remote isn't that straight forward either.